duckduckgo/BrowserServicesKit
# Add support for per-model Sync Data Providers initialization

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/0/1205093266805582/f
iOS PR: https://github.com/duckduckgo/iOS/pull/1863
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1387
What kind of version bump will this require?: Major

**Description**:
Rework Sync initialization mechanism by moving sync data models initialization from
DDGSync (where it was tied to account signup/login flow) to data models themselves.
Sync Metadata is leveraged to store sync setup state per data model. Two values are
possible: _needs remote data fetch_ (a.k.a initial sync) or _ready to sync_.
Decoupling account flow from data models initialization allows to initialize data models
independently of user turning on sync, which is a requirement to support initial sync
for new syncable models added with client app updates.
Also DataProvider class has been added – it's a dedicated sync data providers superclass
that also encapsulates common logic, making individual data providers simpler.



**Steps to test this PR**:
iOS:
1. Start off with a fresh app from App Store. Enable sync and create a few bookmarks.
2. Enter airplane mode.
3. Add new bookmark.
4. Kill the app. Disable airplane mode.
5. Run the app from Xcode.
6. Observe logs and make sure that PATCH request has been sent (and not GET, which would mean that initial sync
  was triggered for an active account).

macOS:
1. Start off with a fresh app from Xcode. Add a bookmark.
2. Enable Sync logging.
3. Join iOS sync account and make sure GET request was sent before a PATCH request.
4. Disable sync and re-enable sync (new sync account). Make sure that PATCH request was sent.
5. Disable sync and re-join iOS account. Make sure that GET request was sent before a PATCH request.



**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
